typing/ctype.ml

```diff
@@ -1305,73 +1300,71 @@ let instance_class params cty =
 
 (**** Instantiation for types with free universal variables ****)
 
-let rec diff_list l1 l2 =
-  if l1 == l2 then [] else
-  match l1 with [] -> invalid_arg "Ctype.diff_list"
-  | a :: l1 -> a :: diff_list l1 l2
-
-let conflicts free bound =
-  let bound = List.map get_id bound in
-  TypeSet.exists (fun t -> List.memq (get_id t) bound) free
-
 let delayed_copy = ref []
     (* copying to do later *)
-
-(* Copy without sharing until there are no free univars left *)
-(* all free univars must be included in [visited]            *)
-let rec copy_sep ~copy_scope ~fixed ~free ~bound ~may_share
-    (visited : (int * (type_expr * type_expr list)) list) (ty : type_expr) =
+let add_delayed_copy t copy_scope ty =
+  delayed_copy :=
+    lazy (Transient_expr.set_stub_desc t (Tlink (copy copy_scope ty))) ::
+    !delayed_copy
+
+(* [copy_sep] is used to instantiate first-class polymorphic types.
+   * It first makes a separate copy of the type as a graph, omitting nodes
+     that have no free univars.
+   * In this first pass, [visited] is used as a mapping for previously visited
+     nodes, and must already contain all the free univars in [ty].
+   * The remaining (univar-closed) parts of the type are then instantiated
+     with [copy] using a common [copy_scope].
+   The reason to work in two passes lies in recursive types such as:
+     [let h (x : < m : 'a. < n : 'a; p : 'b > > as 'b) = x#m]
+   The type of [x#m] should be:
+     [ < n : 'c; p : < m : 'a. < n : 'a; p : 'b > > as 'b > ]
+   I.e., the universal type variable ['a] is both instantiated as a fresh
+   type variable ['c] when outside of its binder, and kept as universal
+   when under its binder.
+   Assumption: in the first call to [copy_sep], all the free univars should
+   be bound by the same [Tpoly] node. This guarantees that they are only
+   bound when under this [Tpoly] node, which has no free univars, and as
+   such is not part of the separate copy. In turn, this allows the separate
+   copy to keep the sharing of the original type without breaking its
+   binding structure.
```

Comment:
The story about using only 2 substitutions is correct. This is what this comment is supposed to say (in terms instantiation rather than substitution, but this means the same thing).

However, it is not true that traversing a [Tpoly] node that binds separate univars does not extend the substitution. If this [Tpoly] is not univar-closed, we need to copy it, and as result we also need to copy its univars. I.e. we extend the substitution with mapping from these univars to fresh univars. So when we duplicate a [Tpoly] node because it contains free univars, we also need to duplicate its whole skeleton (i.e. all the occurrences of the univars bound bind this [Tpoly]), which in turn may force use to duplicate more [Tpoly]'s... This is all fine because we duplicate as a graph.

The last sentence says that, despite the fact the separate copy may contain recursions on `Tpoly`'s, since the only really intantiated univars (to normal variables) are the ones that were given originally, and they are all bound by te same [Tpoly], this separate copy preserves the binding of univars of the original type. Maybe the reason you couldn't understand is just that you didn't see that we were copy all the univars, not just the instantiated ones.

Reply:
Thanks! I understand better now, but the last sentence of your comment is still obscure in my opinion -- I don't think that other people would necessarily understand what you mean.

I would like to understand this algorithm as a copy function with several distinct memoization tables (that let us preserve sharing and provides termination on cycles), and a natural/correct/clear criterion for when to move from one memo table to another. I understand the logic in the current implementation, which is to use two memo tables, one in the traversal contexts where the original-Tpoly-univars are free and another when they are bound. But why is this correct?

My original intuition is that we use a different memo-table for each substitution (from a list of free univars to types). But as you just pointed out, this is not quite right: when we go under a Tpoly node, we extend the current substitution (or, in the case of recursive occurrences of the original Tpoly, we shadow the substitution), so we get a different substitution. But we don't use a different memo table at this point (unless we shadowed the original univars). You speak of "the really instantiated univars" as the memoization key (in addition to the node), so to speak, but again, is there a simple explanation for why this is correct?

(My attempt at reasoning this out below.)

---

If we have a substitution ` rho` and another `rho'` that extends `rho` (in particular, it coincides with `rho` on all univars bound by `rho'`), it is correct to share the same memo-table between the two, under the assumption that all free univars in the substituted term are bound in `rho` (this is the natural extension of your "no other free univar at the beginning" assumption). Indeed, for the nodes that were substituted with `rho`, the univars of `rho'` were not free, so it would be equivalent to substitute them with `rho'`, and so we can consider that all nodes are substituted with `rho'` and may share a single memo-table.

But then we would be in trouble if there was another extension `rho''` of `rho`, in another part of the type graph, that was incompatible with `rho'` because it would bind some of the same univars to different values. This corresponds to a case of two distinct `Tpoly` nodes binding the same univars. Can this happen? (It would probably be fine in this case to share the substitutions of the common univars between the two occurrences, guaranteeing that `rho'` and `rho''` are in fact compatible.)

So as things stands, my explanation would be as follows:
- let's call "initial univars" the univars of the instantiated Tpoly nodee and "initial substitution" the corresponding substitution
- whenever we go under a Tpoly node that does not shadow the initial univars, we can extend the initial substitution in a way that is compatible with all other extensions
- therefore we can use a single memo-table for the initial substitution and all its extensions
- we assume that the only Tpoly nodes that shadow some initial univars is the initial Tpoly node: (1) it shadows all of them at once, and (2) no other univars (from any extension) is free under immediately it
- therefore, when encountering recursive occurrences of the initial Tpoly node, we have an incompatible substitution on the initial univars only; this requires a separate memo-table, which is done by `copy`


---
Question: if I read the code correctly, `copy` does not freshen the univars bound by the `Tpoly` node it encounters (`copy` itself has no Tpoly-specific logic, and `Btype.copy_type_desc` keeps univars unchanged). In particular, separate copies will keep using the same univars, we are not trying to preserve unicity of bound variables. Why can't we use this strategy for `copy_sep`, keep unchanged the univars of the non-initial Tpoly node we encounter during the traversal? Then we would really have only two substitutions, so the explanation would be quite a bit simpler.

Reply:
For the reasoning, this is indeed what I had in mind. I agree that the last sentence is too short to express what you explained in 20 lines, but should we really put proofs (even informal) in the code, and if not there, where ?

For the question, there is a misunderstanding, as for `Tunivar` as for `Tvar` the identity of the type is that of the node. So even if `copy_type_desc` is the identity we do get a fresh type node.
This said, I have also simplified `copy_sep`, to remove the `Tpoly` case that appears to be no longer necessary. As a side-effect, this removes the reliance on the invariant that the same `Tunivar` cannot be bound by physically distinct (but semantically identical) `Tpoly`.

Reply:
Nice simplification.

I think that the simplified version supports my intuition that we are not extending the substitution when encountering Tpoly node whose univars are disjoint from the domain. In a paper definition we would write something like

```
x \notin \rho
--------------------------------
(lam x. t)[rho] := lam x.t[rho]
```

Note that the `x` in the result is a different binder from the `x` in the input, so we implicitly freshen as in `copy_sep`. We could write `[rho, x/x]` on the right-hand-side but it is not necessary.

So we could justify this function with a model with only two distinct substitutions, without considering extensions.

Reply:
That works only if you assume that we track scoping properly. Currently, we rather follow the Barendregt convention: bound type variables occurring in different places must be different. This makes reasoning easier.

Reply:
I don't agree/understand. My point is that we perform a substitution on the free occurrences of the initial univars of the instantiation, and a *copy* (in a memo-table that is indexed by the substitution) on all other nodes, including the other univars (so: no extension of substitutions). There are two different subtitutions, the instantiating substitution we start from, and the empty substitution for subterms that do not contain free occurrences of the initial univars (including under the recursive occurrences of the initial Tpoly node).

This sounds like a simple specification of this function (much simpler than considering compatible extensions etc.), and it does clearly correspond to the code now that you have removed the specific Tpoly handling.

Reply:
> should we really put proofs (even informal) in the code, and if not there, where ?

To understand functions on OCaml types, I need two things:

1. A high-level specification of what the function does, as an operation on type-graphs, described as declaratively as possible. (This high-level operation should make sense to me.)
2. An explanation of how that operation is realized in the implementation I am about to read. (The relation between the abstract operation and the implementation choices should make sense to me.)

The high-level operation is a substitution with substitution-indexed memo tables. (The memo table matter in the high-level specification as they are necessary for termination, even if we don't care too much about preserving sharing in general.) The discussion we are having (just two substitutions, or their many extensions under additional free univars) is about (1) how to define/present this high-level operation. I understand both choices of presentation. But one makes (2) very clear (there are two substitutions in the abstract, so two memo table, and I see how they are realized in the implementaiton), and the other (with extensions) makes (2) much less clear, because I would then naturally expect the implementation to have many memo-tables. In this second case, to understand (2), the relation between the specification and the implementation, I would indeed need an informal correctness argument for why many memo-tables can be condensed in just two. This is just about making sense of the code (and it's a relation to the specification), not a general belief that we should always informally prove correctness.

Reply:
Should we schedule another online discussion?
I have been spending more time writing explanations here than fixing the bug. If I need to write a paper on the subject before we fix the bug, this all becomes extremely time consuming, and I will end up not fixing bugs.

I'm not sure of what would qualify as high-level specification for you.
I repeat what this function is doing at the graph level:
* it first produces a copy of the part of the graph reachable from the root that contain free univars; in this copy the univars we are instantiating are replaced with normal variables, the other ones are replaced by fresh univars (because when you copy a `Tunivar` you are creating a fresh one). This copying ignores levels, as it is required to preserve the binding structure.
* it then connects this copy with another instance of the graph corresponding to parts that do not contain free univars, and as such can use the normal `copy` function, whose behavior depends on levels.
* there are some subtle cases for objects and variants, but they are better understood separately.

Reply:
OK, let me try to give a theoretical specification. This will ignore objects and polymorphic variants for now.

Let's start with trees rather than graphs.
The idea is indeed that we are copying with a substitution for univars. This substitution can return either a normal variable or a fresh univar. I right `'A` for univars.
```
t[rho] = t       if FU(t) = emptyset
'A[rho] = rho('A)
Tpoly(t, 'A1...'An)[rho] = Tpoly(t[rho,'B1...'Bn/'A1...'An], 'B1...'Bn)
other cases are stuctural
```
where `FU(t)` stands for the free univars of `t`.
We can simulate the copying on the remainder by creating the fresh nodes at generic level and taking an instance of `t[rho]` with `copy`, so that part is orthogonal.

Now, of course this specification does not work on cyclic graphs. When can we share the result of substitution?
Basically we can have `t[rho] = t[rho']` if `rho|FU(t) = rho'|FU(t)`.
Without any assumption, this can be tricky, and gave the criterion in the original code.
But if we assume that the original `rho` contains only univars bound in the same `Tpoly`, and that any univar is only bound by one `Tpoly` node, then the scoping rules tell us that whenever we apply the substitution on the same node, we have indeed `rho|FU(t) = rho'|FU(t)`.
* the assumption tells us that the root `Tpoly` node has no free univars, so that `t_root[rho] = t_root[rho'] = t_root` by the first rule. This also means that we don't overwrite those univars in `rho`.
* for other nodes, when we reach a shared node, scoping tells us that the only univars that can be free in it are those that were bound on all the branches that lead to it, or those in the original `rho`. But then, since we assumed that a univar is only bound by a single `Tpoly` node, this means that this node was already shared, and that all the univars it binds are mapped to the same fresh univars in `rho` and `rho'`. Similarly, those from the original `rho` always map to the same variable. Hence `rho|FU(t) = rho'|FU(t)`, and we can share the result of substituting this node.

So it is fine to preserve all sharing. And once we preserve all sharing, the polymorphic variant and object cases become reasonably easy.